src: improve node::Dotenv trimming

dario-piotrowicz opened the pull request:
the trimming functionality that the dotenv parsing uses currently only takes into consideration plain spaces (' '), other type of space characters such as tabs and newlines are not trimmed, this can cause subtle bugs, so the changes here make sure that such characters get trimmed as well

Fixes #56686 

<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

anonrig commented on src/node_dotenv.cc:
```diff
@@ -107,13 +107,28 @@ Local<Object> Dotenv::ToObject(Environment* env) const {
-  if (input.front() == ' ') {
-    input.remove_prefix(input.find_first_not_of(' '));
+
+  size_t input_size = input.size();
+
+  size_t pos_start = 0;
+  while (pos_start < input_size) {
+    const char ch = input.at(pos_start);
+    if (ch != ' ' && ch != '\t' && ch != '\n') {
+      break;
+    };
+    pos_start++;
   }
```
You can do the same thing for other while loop as well:

```suggestion
  auto pos_start = input.find_first_not_of(" \t\n");
  if (pos_start == std::string_view::npos) {
    input.clear();
    return;
  }
```

anonrig commented on test/parallel/test-dotenv-edge-cases.js:
```diff
@@ -137,6 +137,29 @@ describe('.env supports edge cases', () => {
+    // Ref: https://github.com/nodejs/node/issues/56686
+    const code = `
+      process.loadEnvFile('./lines-with-only-spaces.env');
+      console.log(\`EMPTY_LINE: "\${process.env.EMPTY_LINE}"\`);
+      console.log(\`SPACES_LINE: "\${process.env.SPACES_LINE}"\`);
+      console.log(\`TABS_LINE: "\${process.env.TABS_LINE}"\`);
+      console.log(\`SPACES_TABS_LINE: "\${process.env.SPACES_TABS_LINE}"\`);
```
Can you replace them with assert's just like in line 127

dario-piotrowicz replied:
yes, I should have done that for consistency already 😓 , although I thought it's be cleaner/clearer to have the asset in the actual js code instead of in the code string, is there any specific benefit in that? 🤔

anonrig commented on src/node_dotenv.cc:
```diff
@@ -107,13 +107,28 @@ Local<Object> Dotenv::ToObject(Environment* env) const {
   if (input.empty()) return "";
-  if (input.front() == ' ') {
-    input.remove_prefix(input.find_first_not_of(' '));
+
```
Can you add a comment to what this function does after your changes?

dario-piotrowicz replied:
sure 👍

dario-piotrowicz replied:
added, please let me know if the comment works for you 🙂

anonrig commented on src/node_dotenv.cc:
```diff
@@ -105,15 +105,22 @@ Local<Object> Dotenv::ToObject(Environment* env) const {
-  if (!input.empty() && input.back() == ' ') {
-    input = input.substr(0, input.find_last_not_of(' ') + 1);
+
+  size_t pos_end = input.find_last_not_of(" \t\n");
```
Technically it returns a std::string_view::iterator. Let's use auto to avoid making this line longer...

```suggestion
  auto pos_end = input.find_last_not_of(" \t\n");
```

anonrig commented on src/node_dotenv.cc:
```diff
@@ -105,15 +105,22 @@ Local<Object> Dotenv::ToObject(Environment* env) const {
   }
-  return input;
+
+  return input.substr(pos_start, pos_end - pos_start + 1);
```
Please also check for `pos_start == pos_end` before returning substring

dario-piotrowicz replied:
check added 🙂

anonrig commented on src/node_dotenv.cc:
```diff
@@ -105,15 +105,22 @@ Local<Object> Dotenv::ToObject(Environment* env) const {
+
+  size_t pos_end = input.find_last_not_of(" \t\n");
+  if (pos_end == std::string_view::npos) {
+    pos_end = input.size() - 1;
```
If input.size() == 0, this will segfault. If pos_end == std::string_view::npos, you can just return the following:

```suggestion
    return input.substr(pos_start);
```

anonrig commented on test/parallel/test-dotenv-edge-cases.js:
```diff
@@ -137,6 +137,25 @@ describe('.env supports edge cases', () => {
+    // Ref: https://github.com/nodejs/node/issues/56686
+    const code = `
+      process.loadEnvFile('./lines-with-only-spaces.env');
+      require('assert').strictEqual(process.env.EMPTY_LINE, 'value after an empty line');
```
```suggestion
      assert.strictEqual(process.env.EMPTY_LINE, 'value after an empty line');
```

anonrig replied:
I recommend fixing the other tests in this file on a different pull-request.

dario-piotrowicz replied:
I'm also happy to open a follow up PR to update the other tests 🙂👍

anonrig commented:
Looking good. After these changes, we can land this pull-request. @dario-piotrowicz

dario-piotrowicz commented:
@anonrig thank you very much for your review and help here 🙏  (and sorry for the silly mistakes 😓)

I've applied all your suggested changes 🙂

anonrig commented:
@dario-piotrowicz the tests are failing

dario-piotrowicz commented:
> @dario-piotrowicz the tests are failing

Sorry about that, I've addressed it 🙂  (https://github.com/nodejs/node/pull/56983/commits/e3ee074d869ecb6f15a9441fa4436a779662ad98)

The issue was this new check:
```cpp
if (pos_start == pos_end) {
  return "";
}
```

since if `pos_start` is the same as `pos_end` that shouldn't result in an empty string but in a single character at that position (`input.at(post_start)`)

I removed the check completely (instead of returning `input.at(pos_start)`) since when `pos_start` and `post_end` are the same `return input.substr(pos_start, pos_end - pos_start + 1);` should work just fine (as it would resolve to `return input.substr(pos_start, 1);` meaning that it'd take a single character starting from `pos_start`)

do you agree?

jasnell commented on src/node_dotenv.cc:
```diff
@@ -105,15 +105,22 @@ Local<Object> Dotenv::ToObject(Environment* env) const {
-  if (input.front() == ' ') {
-    input.remove_prefix(input.find_first_not_of(' '));
+
+  auto pos_start = input.find_first_not_of(" \t\n");
```
should this also include `\r` and `\f`?

dario-piotrowicz replied:
sounds good to me 🙂, but I'm not sure how to test that, would you be ok with addition without dedicated tests for it?
